src/Microsoft.Health.Fhir.Shared.Core/Features/Resources/Patch/PatchResourceHandler.cs

```diff
@@ -0,0 +1,109 @@
+            Lazy<IConformanceProvider> conformanceProvider,
+            IResourceWrapperFactory resourceWrapperFactory,
+            ResourceIdProvider resourceIdProvider,
+            IAuthorizationService<DataActions> authorizationService,
+            IModelInfoProvider modelInfoProvider)
+            : base(fhirDataStore, conformanceProvider, resourceWrapperFactory, resourceIdProvider, authorizationService)
+        {
+            EnsureArg.IsNotNull(modelInfoProvider, nameof(modelInfoProvider));
+
+            _modelInfoProvider = modelInfoProvider;
+        }
+
+        public async Task<PatchResourceResponse> Handle(PatchResourceRequest request, CancellationToken cancellationToken)
+        {
+            EnsureArg.IsNotNull(request, nameof(request));
+
+            if (await AuthorizationService.CheckAccess(DataActions.Write, cancellationToken) != DataActions.Write)
+            {
+                throw new UnauthorizedFhirActionException();
+            }
+
+            ResourceKey key = request.ResourceKey;
+
+            if (!string.IsNullOrEmpty(key.VersionId))
+            {
+                throw new MethodNotAllowedException(Core.Resources.DeleteVersionNotAllowed);
+            }
+
+            ResourceWrapper currentDoc = await FhirDataStore.GetAsync(key, cancellationToken);
+
+            if (currentDoc == null)
+            {
+                throw new ResourceNotFoundException(string.Format(Core.Resources.ResourceNotFoundById, key.ResourceType, key.Id));
+            }
+
+            if (currentDoc.IsHistory)
+            {
+                throw new MethodNotAllowedException(Core.Resources.PatchVersionNotAllowed);
+            }
+
+            try
+            {
+                ISourceNode node = FhirJsonNode.Parse(currentDoc.RawResource.Data);
+                Resource resource = node.ToTypedElement(_modelInfoProvider.StructureDefinitionSummaryProvider).ToPoco<Resource>();
+                JObject nodeJson = node.ToJObject();
+                string narrative = resource.Scalar(KnownFhirPaths.ResourceNarrative) as string;
+                string narrativeStatus = (string)nodeJson["text"]["status"];
+
+                request.PatchDocument.ApplyTo(nodeJson);
+
+                FhirJsonNode nodePatch = (FhirJsonNode)FhirJsonNode.Create(nodeJson);
+                Resource resourcePatch = nodePatch.ToTypedElement(_modelInfoProvider.StructureDefinitionSummaryProvider).ToPoco<Resource>();
+                string narrativePatch = resourcePatch.Scalar(KnownFhirPaths.ResourceNarrative) as string;
+                string narrativeStatusPatch = (string)nodeJson["text"]["status"];
+
+                if (resource.Id != resourcePatch.Id ||
+                    resource.VersionId != resourcePatch.VersionId ||
+                    resource.Meta.LastUpdated.Value != resourcePatch.Meta.LastUpdated.Value ||
+                    narrative != narrativePatch ||
+                    narrativeStatus != narrativeStatusPatch)
```

Comment:
@brendankowitz the validation is getting long here. If there's a better place/approach for this, let me know. Also let me know if there's a better way to access the narrative status for validation.